File: leetcode/39-combination-sum/solution.py
"""
LeetCode 39. Combination Sum
https://leetcode.com/problems/combination-sum/

题目描述:
给定一个无重复元素的整数数组 candidates 和一个目标整数 target,
找出 candidates 中可以使数字和为目标数 target 的所有不同组合。
同一个数字可以被无限制重复选取。组合内元素的顺序无关。
"""
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        """
        解法:回溯 + 剪枝
        核心思路:对 candidates 排序后,做「选 / 不选」式的 DFS。
        排序的目的:在循环里一旦发现当前累计和 + 候选 > target,
        就可以直接 break,后面的候选更大,更不可能满足。
        """
        # 排序便于剪枝(后面有详细解释)
        candidates.sort()
        res: List[List[int]] = []
        path: List[int] = []

        def dfs(start: int, remain: int) -> None:
            # start: 本次搜索从 candidates[start] 开始选,避免重复组合
            # remain: 还差多少凑出 target
            if remain == 0:
                # 找到一个合法组合,拷贝一份加入结果
                res.append(path.copy())
                return
            # 关键剪枝:remain < candidates[start] 时,再往后只会更大,直接返回
            # (因为数组已排序)
            if remain < candidates[start]:
                return
            for i in range(start, len(candidates)):
                c = candidates[i]
                # 剪枝:加入 c 后超过 target,后面的更大,直接结束本层循环
                if c > remain:
                    break
                # 做选择:把 c 加入路径
                path.append(c)
                # 注意:递归时仍传 i(不是 i+1),允许重复使用同一个数
                dfs(i, remain - c)
                # 撤销选择:回溯到上一层继续尝试其他分支
                path.pop()

        dfs(0, target)
        return res


if __name__ == "__main__":
    sol = Solution()

    # 测试 1:官方例 1
    # candidates = [2,3,6,7], target = 7
    # 期望:[[2,2,3],[7]]
    out1 = sol.combinationSum([2, 3, 6, 7], 7)
    assert sorted(out1) == sorted([[2, 2, 3], [7]]), f"test1 failed: {out1}"
    print(f"test1 passed: {out1}")

    # 测试 2:官方例 2
    # candidates = [2,3,5], target = 8
    # 期望:[[2,2,2,2],[2,3,3],[3,5]]
    out2 = sol.combinationSum([2, 3, 5], 8)
    assert sorted(out2) == sorted([[2, 2, 2, 2], [2, 3, 3], [3, 5]]), f"test2 failed: {out2}"
    print(f"test2 passed: {out2}")

    # 测试 3:官方例 3(target 小于最小候选)
    # candidates = [2], target = 1
    # 期望:[]
    out3 = sol.combinationSum([2], 1)
    assert out3 == [], f"test3 failed: {out3}"
    print(f"test3 passed: {out3}")

    # 测试 4:target = 0
    # 任何 candidates,空组合都满足(和为 0)
    out4 = sol.combinationSum([2, 3, 5], 0)
    assert out4 == [[]], f"test4 failed: {out4}"
    print(f"test4 passed: {out4}")

    # 测试 5:candidates 含较大数,部分组合需要多次重复
    # candidates = [3,4,5], target = 16
    # 期望(无序比较):[[3,3,3,3,4],[3,3,5,5],[4,4,4,4],[3,4,4,5]]
    out5 = sol.combinationSum([3, 4, 5], 16)
    expected5 = sorted([
        [3, 3, 3, 3, 4],
        [3, 3, 5, 5],
        [4, 4, 4, 4],
        [3, 4, 4, 5],
    ])
    assert sorted(out5) == expected5, f"test5 failed: {out5}"
    print(f"test5 passed: {out5}")

    # 测试 6:无重复元素的同一组合不应该被多次收录
    # 比如 candidates=[2,3],target=5 → 只有 [2,3],不应出现两次
    out6 = sol.combinationSum([2, 3], 5)
    assert sorted(out6) == [[2, 3]], f"test6 failed: {out6}"
    print(f"test6 passed: {out6}")

    print("\n所有测试通过 ✓")


# 复杂度分析
# 时间复杂度:O(N^(T/M) + N),其中 N = len(candidates),T = target,
#   M = min(candidates)。最坏情况下每个数都选 M,递归深度 T/M;
#   排序 O(N log N)。
#   题目保证解的总数 < 150,所以实测很快。
# 空间复杂度:O(T/M) — 递归栈和路径长度,最坏 target / min(candidate)。

File: leetcode/39-combination-sum/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * LeetCode 39. Combination Sum
 * https://leetcode.com/problems/combination-sum/
 *
 * 给定无重复元素的整数数组 candidates 和目标 target,
 * 找出所有可以使数字和为目标数 target 的不同组合。
 * 同一个数可以被无限制重复选取。
 */
public class Solution {

    // 结果集:每个内部 List 是一种合法组合
    private final List<List<Integer>> res = new ArrayList<>();
    // 当前正在构造的组合路径
    private final List<Integer> path = new ArrayList<>();
    // 排序后的候选数组(便于剪枝)
    private int[] candidates;

    /**
     * 主入口
     *
     * @param candidates 候选数组(无重复元素)
     * @param target     目标和
     * @return 所有合法组合
     */
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        // 排序后,DFS 中只要发现 c > remain 就可以直接 break,提升效率
        Arrays.sort(candidates);
        this.candidates = candidates;
        dfs(0, target);
        return res;
    }

    /**
     * 深度优先搜索
     *
     * @param start  本轮搜索从 candidates[start] 开始(避免重复组合)
     * @param remain 还差多少凑出 target
     */
    private void dfs(int start, int remain) {
        // 终止条件 1:remain == 0,凑出 target,记录当前组合
        if (remain == 0) {
            res.add(new ArrayList<>(path)); // 必须拷贝,否则后续 path 变更会污染结果
            return;
        }
        // 终止条件 2:排序后 candidates[start] 已经 > remain,后续只会更大,直接返回
        // 这里 start 可能 == candidates.length,需要先判
        if (start >= candidates.length || remain < candidates[start]) {
            return;
        }
        // 关键循环:从 start 开始尝试每个候选
        for (int i = start; i < candidates.length; i++) {
            int c = candidates[i];
            // 剪枝:c 已经 > remain,后面的更大,直接结束循环
            if (c > remain) {
                break;
            }
            // 做出选择:把 c 加入当前路径
            path.add(c);
            // 递归:注意仍是 i(允许重复使用同一数字),不是 i + 1
            dfs(i, remain - c);
            // 撤销选择(回溯),继续尝试其他分支
            path.remove(path.size() - 1);
        }
    }

    public static void main(String[] args) {
        // 注意:每个 case 必须新建 Solution,因为 res/path 是实例字段

        // 测试 1:官方例 1
        Solution sol1 = new Solution();
        int[] c1 = {2, 3, 6, 7};
        List<List<Integer>> out1 = sol1.combinationSum(c1, 7);
        // 期望:[[2,2,3],[7]]
        sortEach(out1);
        sortOuter(out1);
        assert "[[2, 2, 3], [7]]".equals(out1.toString()) : "test1 failed: " + out1;
        System.out.println("test1 passed: " + out1);

        // 测试 2:官方例 2
        Solution sol2 = new Solution();
        int[] c2 = {2, 3, 5};
        List<List<Integer>> out2 = sol2.combinationSum(c2, 8);
        sortEach(out2);
        sortOuter(out2);
        assert "[[2, 2, 2, 2], [2, 3, 3], [3, 5]]".equals(out2.toString())
                : "test2 failed: " + out2;
        System.out.println("test2 passed: " + out2);

        // 测试 3:target 小于最小候选
        Solution sol3 = new Solution();
        int[] c3 = {2};
        List<List<Integer>> out3 = sol3.combinationSum(c3, 1);
        assert "[]".equals(out3.toString()) : "test3 failed: " + out3;
        System.out.println("test3 passed: " + out3);

        // 测试 4:大目标 + 较大候选(组合较长)
        Solution sol4 = new Solution();
        int[] c4 = {3, 4, 5};
        List<List<Integer>> out4 = sol4.combinationSum(c4, 16);
        // 期望四种组合(无序比较)
        sortEach(out4);
        sortOuter(out4);
        String expected4 = "[[3, 3, 3, 3, 4], [3, 3, 5, 5], [3, 4, 4, 5], [4, 4, 4, 4]]";
        assert expected4.equals(out4.toString()) : "test4 failed: " + out4;
        System.out.println("test4 passed: " + out4);

        System.out.println("\n所有测试通过 ✓");
    }

    /** 对每个内部 List 排序,便于断言比较(组合内顺序无关) */
    private static void sortEach(List<List<Integer>> lists) {
        for (List<Integer> l : lists) {
            l.sort(Integer::compareTo);
        }
    }

    /** 对外层 List 排序(按 toString 比较),便于断言比较 */
    private static void sortOuter(List<List<Integer>> lists) {
        lists.sort((a, b) -> a.toString().compareTo(b.toString()));
    }
}

/*
 * 复杂度分析
 * 时间复杂度:最坏情况下回溯会枚举所有组合,
 *   递归深度最多 target / min(candidates),总状态数与解的数量级相关。
 *   排序 O(N log N);题目保证组合数 < 150,实测性能良好。
 * 空间复杂度:O(target / min(candidates)) — 递归栈 + path 长度。
 */
